test/testdata/resolver/basic_constant_out_of_order.rb

```diff
@@ -0,0 +1,46 @@
+# enable-out-of-order-reference-checks: true
+# typed: true
+
+module Foo
+  A = X 
+    # ^ error: `Foo::X` referenced before it is defined
+
+  def self.foo(arg:)
+    X # this is ok
+  end
+
+  def self.bar(&blk)
+    X # this is ok
+  end
+
+  foo arg: ->{ X } # this is ok
+
+  bar do
+    X # this is ok
+  end
+
+  class X; end
+
+  B = X # this is ok
+
+  class Bar
+    Foo::Y
+  # ^^^^^^ error: `Foo::Y` referenced before it is defined
+  end
+
+  class X; end
+
+  class Bar
+    Foo.bar do
+      Foo::Y # this is ok
+    end
+  end
+
+  Y = 1
+
+  class Bar
+    Foo::Y # this is ok
+  end
+
+  Y = 2
```

Comment:
it might be worthwhile to add a case for when a constant is defined three times, because the middle def will neither be the first loc in the file, nor the canonical loc on the symbol. (i can imagine a future change that tries to refactor this logic and fails by way of accidentally overwriting something with the middle loc)